File: 0046-permutations/striver_approach.cpp
// Time complexity: O(n*n!), Space complexity: O(n) for occur vector, O(n) for perm and Auxiliary Stack space of O(n)
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> perm;
        vector<vector<int>> perms;
        int size = nums.size();
        vector<bool> occur(size, false);
        compPerms(nums, perms, perm, occur);
        return perms;
    }
private:
    void compPerms(vector<int>& nums, vector<vector<int>>& perms, vector<int>& perm, vector<bool>& occur) {
        if (perm.size() == nums.size()) {
            perms.push_back(perm);
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            if (!occur[i]) {
                occur[i] = true;
                perm.push_back(nums[i]);
                compPerms(nums, perms, perm, occur);
                occur[i] = false;
                perm.pop_back();
            }
        }
    }
};

File: 0046-permutations/striver_approach.py
# TC: O(n!*n), SC: O(n) for storing visit + O(n) for storing perm.
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        all_perms = []
        size = len(nums)
        perm = []
        visit = [False]*size

        self.create_perm(perm, all_perms, nums, visit, size)
        return all_perms

    def create_perm(self, perm, all_perms, nums, visit, size):
        if len(perm) == size:
            perm_cp = copy.copy(perm)
            all_perms.append(perm_cp)
            return 
        else:
            for i in range(0, size):
                if visit[i] == False:
                    perm.append(nums[i])
                    visit[i] = True
                    
                    self.create_perm(
                        perm = perm, 
                        all_perms = all_perms, 
                        nums = nums, 
                        visit = visit, 
                        size = size,
                    )
                    del perm[-1]
                    visit[i] = False
